src/glang/driver.py

```diff
@@ -77,115 +82,229 @@ def __init__(self):
+                    ".opt.ll"
+                )
+            else:
+                optimized_llvm_output = args.output
+            optimized_llvm_output.write_bytes(optimized_ir)
+
+        return LLVM_IR(optimized_ir, self)
+
+
+class GrapheneSource(Stage):
+    def compile(self, args: DriverArguments) -> LLVM_IR:
+        will_emit_llvm = args.emit_llvm or args.emit_everything
+
+        ir = generate_ir_from_source(
+            self.get_file(), args.include_path, debug_compiler=args.debug_compiler
+        )
+        ir_bytes = ir.encode("utf-8")
+
+        if args.emit_llvm_to_stdout:
+            print(ir_bytes.decode("utf-8"))
+
+        if will_emit_llvm:
+            if args.emit_everything or args.output is None:
+                llvm_output = self.get_top_level_source_file().with_suffix(".ll")
+            else:
+                llvm_output = args.output
+            llvm_output.write_bytes(ir_bytes)
+
+        return LLVM_IR(ir_bytes, self)
+
+
+def link_and_output(args: DriverArguments, objects: list[ELF_Binary]) -> None:
+    linker_args: list[str] = []
+
+    if not args.use_crt:
+        linker_args.extend(["--nostdlib", "--static"])
+
+    binary_output = args.output or Path("a.out")
+    run_checked(
+        [
+            getenv("GRAPHENE_LLD_CMD", "ld.lld"),
+            *linker_args,
+            *(obj.get_file() for obj in objects),
+            "-o",
+            binary_output,
+        ],
+        stdin=None,
+    )
+
+
+def bundle_and_output(args: DriverArguments, objects: list[ELF_Binary]) -> None:
+    lib_output = args.output or Path("lib.a")
+    run_checked(
+        [
+            getenv("GRAPHENE_LLVM_AR_CMD", "llvm-ar"),
+            "rcs",
+            lib_output,
+            *(obj.get_file() for obj in objects),
+        ],
+        stdin=None,
+    )
+
+
+def main() -> None:
+    args = DriverArguments().parse_args()
+    will_emit_binary = (
+        not (args.emit_llvm or args.emit_optimized_llvm or args.emit_llvm_to_stdout)
+        or args.emit_everything
+    )
+
+    load_target_config(args.target)
+
+    parent_dir = Path(__file__).parent.resolve()
+    lib_dir = parent_dir / "lib"
+    target_dir = lib_dir / "std" / get_target()
+
+    if not args.nostdlib:
+        args.include_path.append(lib_dir)
+        args.include_path.append(target_dir)
+
+    assert args.optimize != "2", "@DaveDuck321's coward assert!"
+
+    # 1) Graphene frontend
+    sources: list[GrapheneSource] = [
+        GrapheneSource(source, None) for source in args.inputs
+    ]
+    llvm_sources: list[LLVM_IR] = [source.compile(args) for source in sources]
+
+    # 2) Optimize (if enabled)
+    if args.optimize == "0":
+        opt_sources = llvm_sources  # Don't even touch on O0
+    else:
+        opt_sources = [llvm.optimize(args) for llvm in llvm_sources]
+
+    # 3) Compile
+    if not will_emit_binary:
+        return  # Already done
+
+    objects = [source.compile(args) for source in opt_sources]
```

Comment:
Here we're running the optimizer twice
- `llvm.optimize(...)`
- `source.compile(...)`

It's not a big deal but we can probably save some time compiling e.g. the parse if we only ran `llvm.optimize(...)` when `will_emit_binary == False`

Reply:
This is a bit counter-intuitive, both `llc` and `opt` optimize but they're largely a different set of passes.
As a rule `llc` does all the target-specific stuff + some basic transforms but doesn't do any of the heavy-duty stuff like inlining and dead code elimination. We need both.

Reply:
Ok, that makes sense